### PR TITLE
Issue #8765: add RecordComponentNameCheck

### LINKED ISSUE #8765: Create new check RecordComponentName
From #8764:

```
➜  src /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java                               
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  src cat config.xml                                                                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="LocalVariableName">
            <property name="format" value="(?i)^(?!(record|yield|var)$).+$"/>
        </module>
        <module name="TypeName">
            <property name="format" value="(?i)^(?!(record|yield|var)$).+$"/>
        </module>
    </module>
</module>%                                                                                                                     ➜  src cat TestClass.java                                                                                               
public class TestClass {
    public static void main(String... args) {
        var var = 4; // violation, "var" should not be used as an identifier.

        int record = 15; // violation, "record" should not be used as an identifier.

        String yield = "yield"; // violation, "yield" should not be used as an identifier.

        record Record // violation, "Record" should not be used as an idenitifier.
                (Record record) { // should be violation, "record" should not be used as an identifier.
        }

        String yieldString = "yieldString"; // ok, part of another word
        record MyRecord(){} // ok, part of another word
        var variable = 2; // ok, part of another word
    }
}
➜  src java -jar /home/nick/IdeaProjects/checkstyle/target/checkstyle-8.36-SNAPSHOT-all.jar -c config.xml TestClass.java\

Starting audit...
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:3:13: Name 'var' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [LocalVariableName]
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:5:13: Name 'record' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [LocalVariableName]
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:7:16: Name 'yield' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [LocalVariableName]
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:9:16: Name 'Record' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [TypeName]
Audit done.
Checkstyle ends with 4 errors.

```
I would expect there to be a check that allows users to enfore naming conventions on record components; there should be an ability to specify a `format` property, similar to `MemberNameCheck` (after all, record components are essentially member variables).  Simply adding support for record components to `MemberNameCheck` is not a good solution, since there are many other properties in `MemberNameCheck` that do not apply here. 

Do we need another new check, `RecordComponentName`?

We have LambdaParameterName, MemberName, and  ParameterName.  Since record components are basically "members", we should make it possible for users to enforce naming conventions on them.

### PR DESCRIPTION
**Pull request by strkkk**
Closes #8765

Contrib repo PR: https://github.com/checkstyle/contribution/pull/537

Diff Regression projects: https://gist.githubusercontent.com/strkkk/5eb3b0069ecd438af73bf045c24a855c/raw/d568ee6c30a91e3f721c37b4883ef3264b6203b7/all-projects
New module config: https://gist.githubusercontent.com/strkkk/bf10f326cacdf50f8cb67f0023cd972b/raw/1d80f17c7fb7157e313b393dce0270da2ab01453/lambda_body.xml

### CODE REVIEW

**Comment by strkkk** [comment]
Github, generate report

**Comment by strkkk** [comment]
semaphore failed because of https://github.com/google/error-prone/issues/2092
wercker failure is weird - stackoverflow on
```
Exception in thread "main" java.lang.Error: Error was thrown while processing /pipeline/source/.ci-temp/contribution/checkstyle-tester/src/main/java/local-checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadNoStackoverflowError.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:319)
```

File: src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
```
Comment (pbludov) [inline src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java:3]: 2001-2021?

**Review by pbludov — APPROVED** [review]

File: src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java (hunk 1 above)
Comment (strkkk) [reply src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java]: thanks, fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheck.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * Default value is {@code "^[a-z][a-zA-Z0-9]*$"}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="RecordComponentName"/&gt;
+ * </pre>
+ * <p>
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheck.java:42]: please share code example of behavior of default config.

**Review by romani — CHANGES_REQUESTED** [review]
it very strange that in report we see exceptions
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8765_record_comp_name_2021092851/reports/diff/checkstyle/index.html#A1
we need to suppress them by filter or create 1st config without Checks to collect all such exceptions in report and diff engine will remove them from final report.
@rnveach , @nmancus1 , FYI , ^^^ 

item:

**Comment by nmancus1** [comment]
> we need to suppress them by filter or create 1st config without Checks to collect all such exceptions in report and diff engine will remove them from final report.

config is missing:
```
    <module name="SuppressionSingleFilter">
        <property name="message" value="Exception occurred while parsing"/>
        <property name="checks" value="Checker"/>
    </module>
```

**Comment by romani** [comment]
@strkkk , please recheck from where you get such config, we need to fix that place.

**Comment by strkkk** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheck.java:50]: done

**Comment by strkkk** [comment]
I reused old config. I added filter and now report is ok.

Thanks @nmancus1

**Review by romani — APPROVED** [review]
ok to merge.

**Comment by romani** [comment]
@nmancus1 , you are in review list, please do review.

**Review by nmancus1 — APPROVED** [review]

**Review by rnveach — APPROVED** [review]

**Pull request merged**